src/GSoft.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenCache.cs

```diff
@@ -0,0 +1,73 @@
+﻿using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+
+namespace GSoft.Extensions.Http.Authentication.ClientCredentialsGrant;
+
+/// <summary>
+/// An implementation of a token cache made of two cache levels.
+/// We use a memory cache (L1) for faster look up, then a distributed cache (L2).
```

Comment:
Is there a max size for those caches? If so, do we handle cases where the max size has been reached?

Reply:
For `IDistributedCache`, you can't set a max size. It all depends on the official underlying Microsoft implementation (SQL Server, Redis, etc.).

For `IMemoryCache`, it is possible to set a maximum in-memory cache size. For instance, [`Microsoft.Identity.Web` uses a max cache size of 500MB](https://github.com/AzureAD/microsoft-identity-web/blob/2.9.0/src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs#L60).

I've hesitated to put a max in-memory size here because `Microsoft.Identity.Web`'s cache entries and ours are not comparable.

* Each `Microsoft.Identity.Web`'s cached entry is quite a big JSON document (tens of KB, or maybe hundreds)
* For us, each cache entry is around less than 1k bytes

So it would take quite a lot of cached tokens to fill the memory, and I expect the cache expiration to be very small. Microsoft's `IMemoryCache` (and `IDistributedCache`) already deals with periodic cleanup of expired entries.

Reply:
The mentions of L1 and L2 caches got me confused with CPU cache which is not the same and much smaller, hence my worries. If the size is automatically managed by the language/library then it is better to leave it that way since it has been thoroughly tested.